components/Chips/src/MDCChipField.m

```diff
@@ -0,0 +1,573 @@
+    CGFloat availableWidth = chipFieldMaxX - currentOriginX;
+    // Check if the chip will fit on the current line.  If it won't fit and the available width
+    // is the maximum width, it won't fit on any line. Put it on the current one and move on.
+    if (chipSize.width > availableWidth &&
+        availableWidth < (chipFieldMaxX - self.contentEdgeInsets.right)) {
+      row++;
+      currentOriginX = self.contentEdgeInsets.left;
+    }
+    CGFloat currentOriginY =
+        self.contentEdgeInsets.top + (row * (self.chipHeight + MDCChipFieldVerticalMargin));
+    CGRect chipFrame =
+        CGRectMake(currentOriginX, currentOriginY, chipSize.width, chipSize.height);
+    [chipFrames addObject:[NSValue valueWithCGRect:chipFrame]];
+    currentOriginX = CGRectGetMaxX(chipFrame) + MDCChipFieldHorizontalMargin;
+  }
+  return [chipFrames copy];
+}
+
+- (CGRect)frameForTextFieldForLastChipFrame:(CGRect)lastChipFrame
+                              chipFieldSize:(CGSize)chipFieldSize {
+  CGFloat textFieldWidth =
+      chipFieldSize.width - self.contentEdgeInsets.left - self.contentEdgeInsets.right;
+  CGFloat textFieldHeight = [self.textField sizeThatFits:chipFieldSize].height;
+  CGFloat originY = lastChipFrame.origin.y + (self.chipHeight - textFieldHeight) / 2.f;
+
+  // If no chip exists, make the text field the full width minus padding.
+  if (CGRectIsEmpty(lastChipFrame)) {
+    // Adjust for the top inset
+    originY += self.contentEdgeInsets.top;
+    return CGRectMake(self.contentEdgeInsets.left, originY, textFieldWidth, textFieldHeight);
+  }
+
+  CGFloat availableWidth = chipFieldSize.width - self.contentEdgeInsets.right -
+      CGRectGetMaxX(lastChipFrame) - MDCChipFieldHorizontalMargin;
+  CGFloat placeholderDesiredWidth = [self placeholderDesiredWidth];
+  if (availableWidth < placeholderDesiredWidth) {
+    // The text field doesn't fit on the line with the last chip.
+    originY += self.chipHeight + MDCChipFieldVerticalMargin;
+    return CGRectMake(self.contentEdgeInsets.left, originY, textFieldWidth, textFieldHeight);
+  }
+
+  return CGRectMake(self.contentEdgeInsets.left, originY, textFieldWidth, textFieldHeight);
+}
+
+- (CGFloat)placeholderDesiredWidth {
+  NSString *placeholder = self.textField.placeholder;
+  if (!self.showPlaceholderWithChips && self.chips.count > 0) {
+    placeholder = nil;
+  }
+  UIFont *placeholderFont = self.textField.placeholderLabel.font;
+  CGRect placeholderDesiredRect =
+      [placeholder boundingRectWithSize:self.bounds.size
+                                options:NSStringDrawingUsesLineFragmentOrigin
+                             attributes:@{ NSFontAttributeName : placeholderFont, }
+                                context:nil];
+  return MAX(CGRectGetWidth(placeholderDesiredRect), self.minTextFieldWidth);
+}
+
+#pragma mark - MDCTextInputPositioningDelegate
+
+- (UIEdgeInsets)textInsets:(UIEdgeInsets)defaultInsets {
+  CGRect lastChipFrame = [self.chips.lastObject frame];
```

Comment:
Super-nit: `frame` is a property, please use dot syntax: `self.chips.lastObject.frame`